hardware_in_the_loop/software/modules/cancontroller.py

```diff
@@ -0,0 +1,133 @@
+import threading
+import can
+import time
+import os
+from typing import Callable, Tuple
+
+class CANController:
+    """High level python object to interface with hardware.
+
+    Used to read CAN messages from roadkill harness.
+
+    https://docs.olinelectricmotorsports.com/display/AE/CAN+Controller
+    """
+    def __init__(self, ecus: dict, can_spec_path: str, real_can: bool=True):
+        """
+        Args:
+            ecus (dict): dict of (str: ECU) pairs
+            can_spec_path (str): path to the can spec file
+            real_can (bool): Whether to use real CAN hardware, or to simulate a bus (for testing)
+        """
+        # bring can hardware online
+        if real_can:
+            os.system("sudo ip link set can0 up type can bitrate 500000 restart-ms 100")  # real hardware
+        else:
+            os.system("sudo ip link add dev vcan0 type vcan")
+            os.system("sudo ip link set vcan0 up")  # virtual hardware
+
+        self.ecus = ecus
+        self.read_dict = {}  # Dictionary to help translate raw can to useful signals
+        self.get_states(can_spec_path)
+
+        # start listening
+        bus_name = "can0"  # set above, see socketcan setup docs
+        bus_type = "socketcan" if real_can else "virtual"
+        can_bus = can.interface.Bus(bus_name, bus_type=bus_type)
+        listener = threading.Thread(target=listen, name="listener", kwargs={"can_bus": can_bus, "callback": self.update_ecu})
+        listener.start()
+
+    def get_states(self, path: str) -> None:
+        """Populate each ECUs `states` dictionary, and self.read_dict
+
+        Args:
+            path (str): Path the the CAN spec file
+        """
+        throttle_states = {
```

Comment:
This is what I imagine the state dictionary for each ECU looking like but I may be missing some stuff

Reply:
Using cantools for this as well. This is no longer relevent.